### PR TITLE
fix(sender): graceful restart

### PR DESCRIPTION
**Pull request by colinlyguo**
### Purpose or design rationale of this PR

This PR adds:
* db status rollback.
* blob tx fees precheck.

Note that:
* use permanent delete (Unscoped) instead of soft delete to prevent database bloat, as repeated SendTransaction failures could write a large number of transactions to the database.
* the db role should have "delete" permission after this PR.

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [x] fix: A bug fix

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [x] Yes

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change

## Summary by CodeRabbit

- **New Features**
	- Updated version number to v4.4.79.
	- Introduced a method to delete pending transactions by transaction hash.

- **Bug Fixes**
	- Enhanced error handling and transaction management, including a rollback mechanism for failed transactions.

- **Improvements**
	- Added validation for gas fees in transaction creation to ensure they meet minimum thresholds.
	- Improved clarity and consistency in method naming for transaction status updates in tests.

### CODE REVIEW

File: rollup/internal/controller/sender/sender.go [hunk 1]
```diff
@@ -232,6 +232,12 @@ func (s *Sender) SendTransaction(contextID string, target *common.Address, data
 	}
 
 	if err := s.client.SendTransaction(s.ctx, signedTx); err != nil {
+		// Delete the transaction from the pending transaction table if it fails to send.
+		if updateErr := s.pendingTransactionOrm.DeletePendingTransactionByTxHash(s.ctx, signedTx.Hash()); updateErr != nil {
```
Comment (Thegaram) [inline rollup/internal/controller/sender/sender.go]: Why not move `s.pendingTransactionOrm.InsertPendingTransaction` after `s.client.SendTransaction` inside `SendTransaction`?

Also, what type of error do you have in mind here?

File: rollup/internal/controller/sender/sender.go [hunk 2]
```diff
@@ -458,6 +464,15 @@ func (s *Sender) createReplacingTransaction(tx *gethTypes.Transaction, baseFee,
 				blobGasFeeCap = maxBlobGasPrice
 			}
 
+			// Check if any fee cap is less than double
+			doubledTipCap := new(big.Int).Mul(originalGasTipCap, big.NewInt(2))
+			doubledFeeCap := new(big.Int).Mul(originalGasFeeCap, big.NewInt(2))
+			doubledBlobFeeCap := new(big.Int).Mul(originalBlobGasFeeCap, big.NewInt(2))
+			if gasTipCap.Cmp(doubledTipCap) < 0 || gasFeeCap.Cmp(doubledFeeCap) < 0 || blobGasFeeCap.Cmp(doubledBlobFeeCap) < 0 {
```
Comment (Thegaram) [inline rollup/internal/controller/sender/sender.go:471]: Are we doubling in every gas bump? I thought that aggressive bumping only applies to blob fee cap

File: rollup/internal/controller/sender/sender.go [hunk 3]
```diff
@@ -608,6 +623,23 @@ func (s *Sender) checkPendingTransaction() {
 			}
 
 			if err := s.client.SendTransaction(s.ctx, newSignedTx); err != nil {
+				// SendTransaction failed, need to rollback the previous database changes
```
Comment (Thegaram) [inline rollup/internal/controller/sender/sender.go:626]: Would it be clearer like this?

```go
if err := s.client.SendTransaction(s.ctx, newSignedTx); err != nil {
    tx.Rollback()
} else {
    tx.Commit()
}
```

**Review by Thegaram — DISMISSED** [review]

File: rollup/internal/controller/sender/sender.go (hunk 2 above)
Comment (colinlyguo) [reply rollup/internal/controller/sender/sender.go:471]: aggressive bumping only applies to blob tx, but blob pool will check if `gas fee cap`, `blob gas fee cap` and `gas tip` are all doubled in the meantime.

File: rollup/internal/controller/sender/sender.go (hunk 3 above)
Comment (colinlyguo) [reply rollup/internal/controller/sender/sender.go:626]: this cannot prevent the service is forcefully shutdown by k8s right after sending the tx but not updating db. which is found by `altlayer`. then the sender will keep resubmitting and fail.

File: rollup/internal/controller/sender/sender.go (hunk 1 above)
Comment (colinlyguo) [reply rollup/internal/controller/sender/sender.go]: > Also, what type of error do you have in mind here?

answered in another comment.

> move s.pendingTransactionOrm.InsertPendingTransaction after s.client.SendTransaction

yes. it's the previous implementation, while it cannot prevent forceful shutdown. but during graceful shutdown it does not have any problems (utilizing Golang's context), so this error never happened in sepolia and mainnet before.

File: rollup/internal/controller/sender/sender.go (hunk 3 above)
Comment (colinlyguo) [reply rollup/internal/controller/sender/sender.go:626]: the fix pr of the bug found by `altlayer`'s testing: https://github.com/scroll-tech/scroll/pull/1564

**Review by jonastheis — DISMISSED** [review]

**Review by georgehao — DISMISSED** [review]

**Review by jonastheis — DISMISSED** [review]

**Review by georgehao — APPROVED** [review]

**Pull request merged**